integration_tests/src/tests/offline_processing.rs

```diff
@@ -0,0 +1,171 @@
+
+const BUNNY_URL: &str =
+    "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";
+
+#[test]
+pub fn offline_processing() -> Result<()> {
+    const OUTPUT_FILE: &str = "/tmp/offline_processing_output.mp4";
+    let mut config = read_config();
+    config.queue_options.ahead_of_time_processing = true;
+    config.queue_options.never_drop_output_frames = true;
+    let instance = CompositorInstance::start(Some(config));
+    let (msg_sender, msg_receiver) = crossbeam_channel::unbounded();
+    start_server_msg_listener(instance.api_port, msg_sender);
+
+    instance.send_request(
+        "input/input_1/register",
+        json!({
+            "type": "mp4",
+            "url":  BUNNY_URL,
+            "required": true
+        }),
+    )?;
+
+    instance.send_request(
+        "output/output_1/register",
+        json!({
+            "type": "mp4",
+            "path": OUTPUT_FILE,
+            "video": {
+                "resolution": {
+                    "width": 640,
+                    "height": 320
+                },
+                "encoder": {
+                    "type": "ffmpeg_h264",
+                    "preset": "ultrafast",
+                },
+                "initial": {
+                    "root": {
+                       "type": "view",
+                       "children": [{
+                            "type": "rescaler",
+                            "child": {
+                                "type": "input_stream",
+                                "input_id": "input_1"
+                            }
+                        }]
+                    }
+                },
+                "send_eos_when": { "all_inputs": true }
+            },
+            "audio": {
+                "encoder": {
+                    "type": "aac",
+                    "channels": "stereo"
+                },
+                "initial": {
+                    "inputs": [{ "input_id": "input_1" }]
+                },
+                "send_eos_when": { "all_inputs": true }
+            }
+        }),
+    )?;
+
+    instance.send_request(
+        "input/input_1/unregister",
+        json!({
+            "schedule_time_ms": 2000
+        }),
+    )?;
+    instance.send_request(
+        "output/output_1/unregister",
+        json!({
+            "schedule_time_ms": 2000
+        }),
+    )?;
+
+    instance.send_request("start", json!({}))?;
+
+    for msg in msg_receiver.iter() {
+        if let tungstenite::Message::Text(msg) = msg {
+            if msg.contains("\"type\":\"OUTPUT_DONE\",\"output_id\":\"output_1\"") {
+                info!("breaking");
+                break;
+            }
+        }
+    }
+
+    Command::new("ffprobe")
+        .args(["-v", "error", OUTPUT_FILE])
+        .output()
+        .map_err(|e| anyhow!("Invalid mp4 file. FFprobe error: {}", e))?;
+
+    Ok(())
+}
+
+pub fn start_server_msg_listener(port: u16, event_sender: Sender<tungstenite::Message>) {
```

Comment:
can you extract websocket helpers to the crate in a way that can be adapted to use in other tests too?

Reply:
I'm not sugesting it but just FYI, as an alternative to the websockets you could subscribe to the events directlly (you will get a channel) 

https://github.com/software-mansion/live-compositor/blob/master/compositor_render/src/event_handler.rs